### PR TITLE
Add header folding to `MimeHeaderEncoder`

### PR DESCRIPTION
**Pull request by cketti**
This fixes a crash reported via Google Play (for K-9 Mail 6.800).

```text
Exception java.lang.RuntimeException: An error occurred while executing doInBackground()
  at android.os.AsyncTask$4.done (AsyncTask.java:415)
  at java.util.concurrent.FutureTask.finishCompletion (FutureTask.java:381)
  at java.util.concurrent.FutureTask.setException (FutureTask.java:250)
  at java.util.concurrent.FutureTask.run (FutureTask.java:269)
  at android.os.AsyncTask$SerialExecutor$1.run (AsyncTask.java:305)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1145)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:644)
  at java.lang.Thread.run (Thread.java:1012)
Caused by java.lang.AssertionError: Invalid header
  at com.fsck.k9.mail.internet.MimeHeader.checkHeader (MimeHeader.kt:109)
  at com.fsck.k9.mail.internet.MimeHeader.requireValidHeader (MimeHeader.kt:90)
  at com.fsck.k9.mail.internet.MimeHeader.addHeader (MimeHeader.kt:31)
  at com.fsck.k9.mail.internet.MimeHeader.setHeader (MimeHeader.kt:44)
  at com.fsck.k9.mail.internet.MimeMessage.setHeader (MimeMessage.java:433)
  at com.fsck.k9.mail.internet.MimeMessage.setSubject (MimeMessage.java:281)
  at com.fsck.k9.message.MessageBuilder.buildHeader (MessageBuilder.java:106)
  at com.fsck.k9.message.MessageBuilder.build (MessageBuilder.java:92)
  at com.fsck.k9.message.SimpleMessageBuilder.buildMessageInternal (SimpleMessageBuilder.java:33)
  at com.fsck.k9.message.MessageBuilder$1.doInBackground (MessageBuilder.java:532)
  at com.fsck.k9.message.MessageBuilder$1.doInBackground (MessageBuilder.java:529)
  at android.os.AsyncTask$3.call (AsyncTask.java:394)
  at java.util.concurrent.FutureTask.run (FutureTask.java:264)
Caused by com.fsck.k9.mail.internet.MimeHeaderParserException:
  at com.fsck.k9.mail.internet.UnstructuredHeaderChecker.checkHeaderValue (MimeHeaderChecker.kt:65)
  at com.fsck.k9.mail.internet.MimeHeaderChecker.checkHeader (MimeHeaderChecker.kt:25)
  at com.fsck.k9.mail.internet.MimeHeader.checkHeader (MimeHeader.kt:106)
```

The problem was that long lines were encoded using the encoded word encoding, but the result was not [folded](https://datatracker.ietf.org/doc/html/rfc5322#section-2.2.3). So the maximum line length (998 characters) could be exceeded.

To avoid sending invalid messages that e.g. exceed the maximum line length we use `MimeHeaderChecker` to catch those errors and crash the app.

### CODE REVIEW

File: mail/common/src/test/java/com/fsck/k9/mail/internet/MimeHeaderEncoderTest.kt
Comment (wmontwe) [inline mail/common/src/test/java/com/fsck/k9/mail/internet/MimeHeaderEncoderTest.kt:1]: There should be a test to also verify the boundary condition at the recommended line lenght to ensure correct behavior and one to check behavior for a long string that does not require folding.

File: mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt [hunk 1]
```diff
@@ -17,12 +19,15 @@ object MimeHeaderEncoder {
     }
 
     private fun exceedsRecommendedLineLength(name: String, value: String): Boolean {
-        return name.length + COLON_PLUS_SPACE_LENGTH + value.length > RECOMMENDED_MAX_LINE_LENGTH
+        return name.headerPrefixLength + value.length > RECOMMENDED_MAX_LINE_LENGTH
```
Comment (wmontwe) [inline mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt:22]: Using an extension function on `String` could lead to misunderstandings. It is not clear that it should only be used for the name and should not be applied to value.

File: mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt [hunk 2]
```diff
@@ -17,12 +19,15 @@ object MimeHeaderEncoder {
...
     }
 
     private fun charactersNeedEncoding(text: String): Boolean {
         return text.any { !it.isVChar() && !it.isWspOrCrlf() }
     }
 
     private const val COLON_PLUS_SPACE_LENGTH = 2
+
+    private val String.headerPrefixLength: Int
+        get() = length + COLON_PLUS_SPACE_LENGTH
```
Comment (wmontwe) [inline mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt:32]: This could help with the above issue:

```suggestion
    /**
    * Calculates the length of the header prefix, which includes the length of the name and a constant space for colon and space.
    */
    private val String.headerPrefixLength: Int
        get() = length + COLON_PLUS_SPACE_LENGTH
```

or replace it with this:

```suggestion
    /**
    * Calculates the length of the header prefix, which includes the length of the name and a constant space for colon and space.
    */
    private fun headerPrefixLength(name: String): Int {
        return name.length + COLON_PLUS_SPACE_LENGTH
    }
```

**Review by wmontwe — COMMENTED** [review]
Added some readability and test findings.

File: mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt (hunk 1 above)
Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt]: > It is not clear that it should only be used for the name property and should not be applied to value.

The same rationale could be applied when it's an argument to a regular function. But I changed the code to get rid of the extension function.

File: mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt (hunk 2 above)
Comment (cketti) [reply mail/common/src/main/java/com/fsck/k9/mail/internet/MimeHeaderEncoder.kt]: I don't think a comment like this adds any value. It's a 1:1 translation from the code to English.

File: mail/common/src/test/java/com/fsck/k9/mail/internet/MimeHeaderEncoderTest.kt
Comment (cketti) [reply mail/common/src/test/java/com/fsck/k9/mail/internet/MimeHeaderEncoderTest.kt:1]: Currently the behavior of `MimeHeaderEncoder` is not what I want it to be. For now the only important part is that the output satisfies `MimeHeaderChecker`. I'll add more detailed tests when I have time to work on `EncoderUtil`.

Comment (wmontwe) [reply mail/common/src/test/java/com/fsck/k9/mail/internet/MimeHeaderEncoderTest.kt:1]: I asked for boundary checks as at first glance a maximum line length of 998 characters and a test for 999 characters seems reasonable to satisfy [RFC 5322, section 2.1.1](https://datatracker.ietf.org/doc/html/rfc5322#section-2.1.1) MUST max line length.

Due to the use of `MimeUtil.fold` and `RECOMMENDED_MAX_LINE_LENGTH` the behavior is different, but there is not test that validates that we also apply the SHOULD be no more than 78 characters condition.

If that is not by accident, it should be tested.

Comment (cketti) [reply mail/common/src/test/java/com/fsck/k9/mail/internet/MimeHeaderEncoderTest.kt:1]: The behavior regarding the recommended line length isn't great right now. But I've now added a test that checks the line lengths of the output when the input exceeds the recommended line length.

**Review by wmontwe — APPROVED** [review]
👍
